### ISSUE TITLE
VSS processing is time consuming

### ISSUE THREAD

**Issue by utkonos** [log2timeline/plaso#2554]
**Description of problem:**

When working on `/dev` as the source and analyzing all VSS, the process slows to a crawl and takes days to complete.

**Command line and arguments:**

`./psteal.py -o json --source /dev/sdb3 -w output.json`

**Source data:**

This is a hard drive. I'm unable to provide data, sorry.

**Plaso version:**

20190429

**Operating system Plaso is running on:**

Ubuntu 19.04, Python 3.7.3

**Installation method:**

Virtual environment
`pip install -r requirements.txt`
`pip install plaso`

Other users have reported similar problems. I'm ready to work with the authors on debugging this problem. What resource monitoring do I need to do to find the bottleneck?

**Comment by joachimmetz** [comment]
1. can you define what slow is for you, in terms of hours
2. can you provide information on how many VSS your disk has
3. can you provide information on how many resources plaso has
4. can you provide the output of `psteal.py --troubles` since you're using an installation method that is not supported

**Comment by utkonos** [comment]
> 1. can you define what slow is for you, in terms of hours

One of the drives with 9 VSS took 5 days. Another with 2 is three days and counting. I've stopped that one to run it with `--troubles`

> 2. can you provide information on how many VSS your disk has, and how many files in the current and each snapshot

9 and 2, file count shortly.

> 3. can you provide information on how many resources plaso has

~1TB free space on the drive
8GB memory (4.9GB available while plaso running)

> 4. can you provide the output of `psteal.py --troubles` since you're using an installation method that is not supported

Running

**Comment by joachimmetz** [comment]
> I've stopped that one to run it with --troubles

There was no need to stop the running process. You could have run this in parallel. Also see: https://plaso.readthedocs.io/en/latest/sources/user/Troubleshooting.html

> 3 can you provide information on how many resources plaso has

@utkonos you'll need to give me more detail to work with. Otherwise I have insufficient information to solve this issue. The more descriptive you can be, the more likely we can find the cause of a bottleneck.

how many workers? what is the average/peak consumption of memory per worker?
how fast is the storage of the source? local USB 2 (or 3), local SATA, network?
how fast is the storage of the output?
how fast is the storage of the temporary files?

**Comment by utkonos** [comment]
> > I've stopped that one to run it with --troubles
> 
> There was no need to stop the running process. You could have run this in parallel. Also see: https://plaso.readthedocs.io/en/latest/sources/user/Troubleshooting.html
> 

`--troubles` fails without output file and source. It appears to just run the whole thing a second time rather than output the data listed in the troubleshooting documentation.

> > 3 can you provide information on how many resources plaso has
> 
> @utkonos you'll need to give me more detail to work with. Otherwise I have insufficient information to solve this issue. The more descriptive you can be, the more likely we can find the cause of a bottleneck.
> 
> how many workers? what is the average/peak consumption of memory per worker?

1 main and 3 workers. 800M-1G for the main process and 500-700M each worker.

> how fast is the storage of the source? local USB 2 (or 3), local SATA, network?

Source: USB3 local

> how fast is the storage of the output?

SATA3 (6G) local

> how fast is the storage of the temporary files?

SATA3 (6G) local

**Comment by joachimmetz** [comment]
> `--troubles` fails without output file and source. It appears to just run the whole thing a second time rather than output the data listed in the troubleshooting documentation.

this is a short coming in some tools, pending fix: TBD

> 1 main and 3 workers. 800M-1G for the main process and 500-700M each worker.

so we calculate a max of 2G with a peak of 3G per worker and 3 - 4G for the main process
do you recall what the overall physical memory consumption was while plaso was running
if a system is under memory pressure, this can slow down processing significantly

another reason for long processing times can be a high number of files

**Comment by utkonos** [comment]
Output of `log2timeline.py --troubles`

```
$ log2timeline.py --troubles
2019-05-27 15:33:48,270 [INFO] (MainProcess) PID:14843 <data_location> Determined data location: /home/user/plaso-20190429/venv/share/plaso
Using Python version 3.7.3 (default, Apr  3 2019, 05:39:12)
[GCC 8.3.0]
 
Path: /home/user/plaso-20190429/venv/bin/log2timeline.py
 
plaso - log2timeline version 20190429
 
Checking availability and versions of dependencies.
[OK]            Crypto version: 2.6.1
[OK]            artifacts version: 20190320
[OK]            bencode
[OK]            biplist
[OK]            certifi version: 2019.03.09
[OK]            chardet version: 3.0.4
[OK]            dateutil version: 2.8.0
[OK]            dfdatetime version: 20190116
[OK]            dfvfs version: 20190421
[OK]            dfwinreg version: 20190329
[OK]            dtfabric version: 20190120
[OK]            elasticsearch version: 7.0.1
[OK]            future version: 0.17.1
[OK]            idna
[OK]            lz4
[OK]            lzma
[OK]            pefile version: 2019.4.18
[OK]            psutil version: 5.6.2
[OK]            pybde version: 20190317
[OK]            pyesedb version: 20181229
[OK]            pyevt version: 20181227
[OK]            pyevtx version: 20181227
[OK]            pyewf version: 20171104
[OK]            pyfsapfs version: 20190510
[OK]            pyfsntfs version: 20190104
[OK]            pyfvde version: 20190104
[OK]            pyfwnt version: 20181227
[OK]            pyfwsi version: 20181227
[OK]            pylnk version: 20181227
[OK]            pymsiecf version: 20181227
[OK]            pyolecf version: 20181231
[OK]            pyparsing version: 2.4.0
[OK]            pyqcow version: 20181227
[OK]            pyregf version: 20190303
[OK]            pyscca version: 20181227
[OK]            pysigscan version: 20190103
[OK]            pysmdev version: 20190315
[OK]            pysmraw version: 20181227
[OK]            pytsk3 version: 20190506
[OK]            pytz
[OK]            pyvhdi version: 20181227
[OK]            pyvmdk version: 20181227
[OK]            pyvshadow version: 20190323
[OK]            pyvslvm version: 20181227
[OK]            requests version: 2.22.0
[OK]            six version: 1.12.0
[OK]            sqlite3
[OK]            urllib3 version: 1.25.3
[OK]            xlsxwriter version: 1.1.8
[OK]            yaml version: 5.1
[OK]            yara version: 3.10.0
[OK]            zmq version: 18.0.1
[OK]            sqlite3 version: 3.27.2
```

**Comment by joachimmetz** [comment]
Side note `pyewf version: 20171104` this version is experimental and not recommended for production usage. However that should not affect this run since you're running from device:

Also `elasticsearch version: 7.0.1` is not fully tested yet https://github.com/log2timeline/plaso/issues/2465

* do you recall what the overall physical memory consumption was while `psteal.py` was running?
* how long does `psteal.py` take for just the current volume?

**Comment by utkonos** [comment]
> Side note `pyewf version: 20171104` this version is experimental and not recommended for production usage. However that should not affect this run since you're running from device:

I will change this to the supported version between runs. Thanks for the info. I would recommend changing the requirements.txt to place an upper bound on the most recent supported version. Like this:
`libewf-python >= 20131210, < 20171104`
I'm not sure which is the most recent supported version, but just switch that actual version with my example.

> Also `elasticsearch version: 7.0.1` is not fully tested yet #2465

I'm not using elasticsearch yet, but again, I suggest putting an upper bound on the last tested version of ES:
`elasticsearch >= 6.0, < 7.0`

> * do you recall what the overall physical memory consumption was while `psteal.py` was running?
> * how long does `psteal.py` take for just the current volume?

```
              total        used        free      shared  buff/cache   available
Mem:          7.7Gi       3.8Gi       152Mi       1.0Gi       3.7Gi       2.6Gi
Swap:         975Mi       165Mi       810Mi
```

`psteal.py` takes about 3-8 hours for the main volume if no VSS are selected for analysis. This depends on the size of the volume. When the current run ends (perhaps 2-3 days) I will re-run it without VSS to make a comparison.

**Comment by joachimmetz** [comment]
> I would recommend changing the requirements.txt to place an upper bound on the most recent supported version. 

thx for the suggestion but no, pypi is **not** a preferred distribution method for non-development versions of plaso. To repeat https://github.com/log2timeline/plaso/wiki/Running-plaso-in-virtualenv
```
Note that this is intended for development use only, and if you aren't comfortable debugging package installation, this is not for you.
```

So for development purposes we don't want to set an upper bound version.

> This depends on the size of the volume.

The byte size of the volume is no significant factor, it is the number of files and the type of files.

**Comment by joachimmetz** [comment]
So `152Mi` of free memory is not much and I see swap is used. If memory is swapped due to high memory usage, this will impact your systems performance. Try profiling memory swap behaviour and see if limiting the number of workers makes a difference. Might be cheaper to add more RAM.

**Comment by utkonos** [comment]
I increased the number of workers to 6. I will re-run this with the default number and post the memory consumption during that run. It may take a moment for this one to end.

**Comment by joachimmetz** [comment]
You likely want to decrease it seeing:
so (by default) we calculate a max of 2G with a peak of 3G per worker and 3 - 4G for the main process

I repeat monitor for swapping, it can slow down processing significantly

Otherwise have a look at using the profilers if you can pinpoint the bottleneck. Also see: https://plaso.readthedocs.io/en/latest/sources/developer/Profiling.html

**Comment by joachimmetz** [comment]
@utkonos any findings from your profiling?

**Comment by utkonos** [comment]
Working on the profiling. I needed to use that workstation for another task. But I'll be able to finish the profiling very soon.

**Comment by joachimmetz** [comment]
Closing issue. Not actionable without profiling data.Reopen once data is available.
